### PR TITLE
Generate both shortcut and regular views for AST desc types

### PR DESCRIPTION
**Pull request by NathanReb**
This modifies `Ppx_ast.Viewer` so that it provides both shortcuts and regular views for a desc types.

E.g. for the `Pexp_constant` constructor it will now generate:
```ocaml
val pexp_constant'const : (constant, 'i, 'o) View.t -> (expression_desc, 'i, 'o) View.t
val econstant'const : (constant, 'i, 'o) View.t -> (expression, 'i, 'o) View.t
```
instead of
```ocaml
val pexp_constant'const : (constant, 'i, 'o) View.t -> (expression, 'i, 'o) View.t
```

Using the full constructor name will give you the regular view, while using the shortened one will give you the shortcut.

You can see how the short names are built [here](https://github.com/ocaml-ppx/ppx/pull/65/files#diff-7ac1c0bfca27b56b8b3135ba6275c084R11). I'm happy to change those if you have a better naming scheme in mind!

### CODE REVIEW

**Comment by NathanReb** [comment]
Just rebased on top of master, CI should go green now!

File: ast/cinaps/gen_viewer.ml [hunk 1]
```diff
@@ -5,6 +5,25 @@ let string_of_ty ty = Grammar.string_of_ty ~internal:false ty
...
+    | "core_type" -> "t"
+    | "structure_item" -> "str"
+    | "signature_item" -> "sig"
+    | "module_expr" -> "me"
+    | "module_type" -> "mt"
+    | "class_field" -> "cf"
+    | "class_expr" -> "ce"
+    | "class_type" -> "ct"
+    | "class_type_field" -> "ctf"
+    | s -> failwith "No prefix for shortcut: " ^ s
```
Comment (ghost) [inline ast/cinaps/gen_viewer.ml:32]: Could this line cause a failure when trunk adds a new outer record type?

Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: Yes it could! It's not the only such line, for instance we have a few `assert false` lying around in places where it's unclear how we should behave, e.g. if we have shortcuts to polymorphic types.

Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: I agree in this case we could probably have a default fallback for the shortcut prefix and fix it to something shorter and more meaningful when we actually commit the new versioned viewers.

How about falling back to `"Sc" ^ s`?

Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: Here is my thinking: such functions are only necessary for when people want to start migrating their ppx to a newer version of the AST. So we could use a model with two parameters:

1. the highest version we can convert to and from, which we be the max between the higest version in the history of the compiler and the local copy in ppx
2. the highest developers of ppx know about. This would one be bumped manually and could match the version of the ast used in the ppx library itself

we could generate such modules only up to (2). We might then get errors when bumping 2, but that's fine as it would be bumped manually. WDYT?

/cc @ceastlund  as well

Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: Yeah I think that's better and more reasonable. We won't be able to generate code for any arbitrary change in the AST anyway.

To sum up, we would generate `conversions` and eventually `versions` and `traverse` as they are relatively simple for up to (1) and `builder` and `viewer` for up to (2). Is that correct?

Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: I would only generate `conversions` up to 1, that's the only thing we need to ensure compatibility with trunk. This way, you'd get full support for developing against a new ast version at once, rather than minimal support followed by full support. Indeed, minimal support would quickly be very limited to do anything serious.

Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: Perfect, we may leave the `failwith` then!

Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: In any case, we can sort that out in a separate PR.

**Review by ghost — APPROVED** [review]
Looks good!

File: ast/cinaps/gen_viewer.ml (hunk 1 above)
Comment (ceastlund) [reply ast/cinaps/gen_viewer.ml:32]: I think they two meaningful options for things like this is either fail on unknown, which means certain AST changes will break ppx for that version and force it to update -- which it needs to do anyway -- or to ignore unknown cases and not generate helpers for them, which means new versions work instantly but may need manual help to make them complete. I suspect in the long run neither approach is perfect and it doesn't make a huge difference.

File: traverse/ppx_traverse.ml [hunk 2]
```diff
@@ -391,7 +391,7 @@ let erase_type_variables = object
   method! core_type core_type =
     let open Viewer in
     match%view core_type with
-    | Ptyp_var _ [@view? {ptyp_loc = loc; _}] -> Builder.ptyp_any ~loc
+    | Tvar _ [@view? {ptyp_loc = loc; _}] -> Builder.ptyp_any ~loc
```
Comment (ceastlund) [inline traverse/ppx_traverse.ml]: Lines like this should be updated to use Ptyp_var and the proper fields, rather than using Tvar and an attribute.

**Review by ceastlund — COMMENTED** [review]
It'd be nice to clean up the places using attributes to use the longhand patterns instead, other than that looks good to go.

File: ast/cinaps/gen_viewer.ml (hunk 1 above)
Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: I think I prefer the failure mode because it won't let things slip through and we'll always have to deal with them.

File: traverse/ppx_traverse.ml (hunk 2 above)
Comment (NathanReb) [reply traverse/ppx_traverse.ml]: I kept the `@view` patterns because it was shorter but I'm happy to switch to the full pattern if you prefer.

File: ast/cinaps/gen_viewer.ml (hunk 1 above)
Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: Having the new versions work instantly but may need manual help has one problem though: in some cases the manual help might mean breaking a bit the API. So if people start using the new functions before we do the manual work, their code might break when we do the manual work.

Comment (NathanReb) [reply ast/cinaps/gen_viewer.ml:32]: How could that happen? We would only release once the manual work is done and the automated changes (`Conversion`) aren't exposed anyway.

Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: I was referring to "to ignore unknown cases and not generate helpers for them, which means new versions work instantly but may need manual help to make them complete" which I interpreted as:

for all versions up to (1), we generate all the modules, using a fail-proof policy that simply skip things we don't know how to name. Then, if things are incomplete or messed up we do a new release of ppx that fixes the generation for the most recent versions of the AST.

So basically it's a best effort policy. It has the advantage that when no manual adjustments are needed, you don't even need to press a button to get the helpers and co for a new version of the AST. But on the other hand, it means that sometimes there will be versions out there of the "ppx" package that will have incomplete APIs, or worse APIs that we don't want at all.

Comment (ceastlund) [reply ast/cinaps/gen_viewer.ml:32]: The only things that can "fail" should be assigning shorthand. And we should always be able to provide that without breaking compatibility by choosing non-overlapping shorthand for any new types.

Comment (ghost) [reply ast/cinaps/gen_viewer.ml:32]: Agreed. I'm just thinking that a new type might make us reconsider some of the existing shorthands. The current shorthands work well for the current set of types, but with a different set of types we might want to use different shorthands altogether. Maybe it's a bit far-fetched, but if we bump (2) manually at least we always have this choice.

**Comment by ceastlund** [comment]
Rebased; merging because this has been sitting a while and I could use it. I think we had pretty much resolved everything.

**Comment by NathanReb** [comment]
Hmm I thought this had been merged. Thanks!
